stan/math/prim/mat/fun/gp_dot_prod_cov.hpp

```diff
@@ -15,142 +16,308 @@ namespace stan {
+
+  for (size_t i = 0; i < x_size; ++i) {
+    x_new[i] = sigma_squared * x[i];
+  }
 
   for (size_t i = 0; i < (x_size - 1); ++i) {
-    cov(i, i) = sigma_sq + dot_self(x[i]);
+    cov(i, i) = dot_product(x[i], x_new[i]);
     for (size_t j = i + 1; j < x_size; ++j) {
-      cov(i, j) = sigma_sq + dot_product(x[i], x[j]);
+      cov(i, j) = dot_product(x[i], x_new[j]);
       cov(j, i) = cov(i, j);
     }
   }
-  cov(x_size - 1, x_size - 1) = sigma_sq + dot_self(x[x_size - 1]);
+  cov(x_size - 1, x_size - 1) = dot_product(x[x_size - 1], x_new[x_size - 1]);
   return cov;
 }
 
 /**
- * Returns a dot product covariance matrix. A member of Stan's Gaussian
- * Process Library.
+ * Returns a dot product covariance matrix of differing
+ * x's. A member of Stan's Gaussian Process Library.
  *
- * \f$k(x,x') = \sigma^2 + x \cdot x'\f$
+ * \f$k(x,x') = x^T sigma\_squared I x'\f$
  *
- * A dot product covariance matrix is the same covariance matrix
- * as in bayesian regression with \f$N(0,1)\f$ priors on regression coefficients
- * and a \f$N(0,\sigma^2)\f$ prior on the constant function. See Rasmussen and
- * Williams et al 2006, Chapter 4.
+ * @tparam T_x1 type of first std::vector of double
+ * @tparam T_x2 type of second std::vector of double
+ * @tparam T_sigma_squared scalar type of sigma_squared
  *
- * @tparam T_x type of std::vector of double
- * @tparam T_sigma type of sigma
+ * @param x1 std::vector of elements that can be used in dot_product
+ * @param x2 std::vector of elements that can be used in dot_product
+ * @param sigma_squared variance
+ * @return dot product covariance matrix
+ * @throw std::domain_error if sigma < 0, nan or inf
+ *   or if x1 or x2 are nan or inf
+ */
+template <typename T_x1, typename T_x2, typename T_sigma_squared>
+Eigen::Matrix<return_type_t<T_x1, T_x2, T_sigma_squared>, Eigen::Dynamic,
+              Eigen::Dynamic>
+gp_dot_prod_cov(const std::vector<Eigen::Matrix<T_x1, Eigen::Dynamic, 1>> &x1,
+                const std::vector<Eigen::Matrix<T_x2, Eigen::Dynamic, 1>> &x2,
+                const T_sigma_squared &sigma_squared) {
+  check_not_nan("gp_dot_prod_cov", "sigma_squared", sigma_squared);
+  check_positive_finite("gp_dot_prod_cov", "sigma_squared", sigma_squared);
+
+  size_t x1_size = x1.size();
+  size_t x2_size = x2.size();
+  for (size_t i = 0; i < x1_size; ++i) {
+    check_not_nan("gp_dot_prod_cov", "x1", x1[i]);
+    check_finite("gp_dot_prod_cov", "x1", x1[i]);
+  }
+  for (size_t i = 0; i < x2_size; ++i) {
+    check_not_nan("gp_dot_prod_cov", "x2", x2[i]);
+    check_finite("gp_dot_prod_cov", "x2", x2[i]);
+  }
+  Eigen::Matrix<return_type_t<T_x1, T_x2, T_sigma_squared>, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x1_size, x2_size);
+
+  if (x1_size == 0 || x2_size == 0) {
+    return cov;
+  }
+
+  check_size_match("gp_dot_prod_cov", "dimension of elements of x1",
+                   size_of(x1[0]), "dimension of elements of x2",
+                   size_of(x2[0]));
```

Comment:
Forgive my ignorance, but where is `size_of` defined? I can't find it in the codebase.